myia/ir.py

```diff
@@ -1,7 +1,142 @@
+returning a nested function creates a closure.
+
+"""
+from enum import Enum, auto
+from typing import List, Union, Dict
+
+LiteralType = Union[str, int, float, 'PrimitiveOperation', 'Graph', None]
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        output: The `Node` whose value will be returned by this function.
+
+    """
+
+    def __init__(self, output: 'Node') -> None:
+        """Construct a graph."""
+        self.output = output
+        self.debug: Dict = {}
 
 
 class Node:
-    """A node in the graph-based ANF IR."""
+    """A node in the graph-based ANF IR.
+
+    There are three types of nodes: Values, which are the result of function
+    applications; parameters; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a value, the first node input is the function to
+            apply, followed by the arguments.
+        value: The value of this node, if it is a constant.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self) -> None:
+        """Construct a node."""
+        self.inputs: List[Node] = []
+        self.value: LiteralType = None
+        self.graph: Graph = None
+        self.debug: Dict = {}
+
+
+class Value(Node):
+    """A value.
+
+    A value is the result of function application.
+
+    """
+
+    def __init__(self, inputs: List[Node], graph: 'Graph') -> None:
+        """Construct a value."""
+        if len(inputs) < 1:
+            raise ValueError
+        super().__init__()
+        self.inputs = inputs
+        self.graph = graph
+
+
+class Parameter(Node):
+    """A parameter to a function.
+
+    Parameters are leaf nodes, since they are not the result of a function
+    application, and they have no value. They are entirely defined by the graph
+    they belong to.
+
+    """
+
+    def __init__(self, graph: Graph) -> None:
+        """Construct the parameter."""
+        super().__init__()
+        self.graph = graph
+
+
+class Constant(Node):
+    """A constant node.
+
+    A constant is a node which is not the result of a function application. In
+    the graph it is a leaf node. It has no inputs, and instead is defined
+    entirely by its value. Unlike parameters and values, constants do not
+    belong to any particular function graph.
+
+    """
+
+    def __init__(self, value: LiteralType) -> None:
+        """Construct a literal."""
+        super().__init__()
+        self.value = value
+
+
+class Literal(Constant):
+    """A literal.
+
+    A literal is a string or a numeric value.
+
+    """
+
+    def __init__(self, value: Union[str, float, int]) -> None:
+        """Construct a literal."""
+        super().__init__(value)
+
+
+class PrimitiveOperation(Enum):
+    """Built-in primitive operations."""
```

Comment:
I don't think this should be an Enum. I think this should be a type, so that it is easy to add primitives, even from outside the library. I'm also not sure it should be in this file.

Reply:
Agreed it probably shouldn't be in this file, but I didn't write any other file yet and wanted the code to at least make sense :)

What is the specific use case for adding primitives from outside the library? I have not seen any compiler representation set up like this. I think part of this is shifting from the TensorFlow/Theano mindset. In that paradigm, "ops" conflate writing adjoints and writing lower-level code. I think that instead primitives should be exactly that: primitive. Supplying custom adjoints for functions can be done using a general "custom adjoint procedure", and calling specific low-level code should be done through an "external function call" interface. This way the runtime, lowering passes, etc. won't choke on new primitives they never saw before, nor do we have an exponential growth in the complexity of the graph that the optimizer needs to deal with.

Reply:
I think compilers are unnecessarily rigid about this, and that this is generally a design flaw in most of them. I don't think a use case is needed either: locking down the set of primitives is not easier than leaving it open, and they generally don't interact with each other in complicated ways, if at all.

> This way the runtime, lowering passes, etc. won't choke on new primitives they never saw before, nor do we have an exponential growth in the complexity of the graph that the optimizer needs to deal with.

All of these should also be easy to extend. I don't know what you mean by "exponential growth in the complexity of the graph" either. These primitives will be opaque to the optimizer, just like an "external function call" would be, unless the user adds optimizations that work with them, which they should be able to do.

Reply:
Happy to change this to classes (we might also want to use the sea-of-nodes approach, where classes of primitives are also the place where e.g. algebraic identities are defined for CP).

Let me clarify my argument though, which is less about the implementation details than about the general approach: Ops do interact with each other e.g. `exp(x)/sum(exp(x)) == exp(x + c)/sum(exp(x + c))`, `mean(x) = sum(x) / len(x)`, etc. The number of possible interactions that you want to consider for optimizations grows as *n<sup>x</sup>* for *n* ops where *x* is the size of the pattern you are trying to match.

So let's say that I implement a new `mean` operation, alongside with an an optimization that says `sum(x) / len(x) -> mean(x)`. However, there already was an optimization that did `sum(x) -> n if x = ones(n)`. Now `sum(x) / len(x) -> n / len(x) -> 1` if the latter optimization gets applied first, but `sum(x) / len(x) -> mean(x)` if the former gets applied first. So now you have to carefully consider how your optimizations interact with the (exponentially growing number of) other optimizations, by either enforcing an order, or by writing a second optimization that does `mean(x) -> 1 if x = ones(n)`.

This is just an illustration of the fact that introducing new primitives should be done with care, because otherwise they can just slow your program down by messing up other optimization passes. This is probably one of the reasons that traditional compilers focus on doing a small number of primitives right. Theano just adds ops for everything, and I think that is one of the reasons that its optimization passes have gotten so messy.

In the end, I'm not proposing any hard restrictions as much as a guideline: Just like I don't want to have to extend LLVM in order to write my C++ code, I don't want users having to write Myia primitives in order for their code to run fast (I'd rather have them write optimizations to detect patterns that can be replaced by kernel calls, which are represented as "external calls" in the IR).